File: api/schemas.py
"""
Pydantic schemas for API request and response bodies.

Why separate schemas from ORM models?
  SQLAlchemy ORM models are tied to the database — they have relationship objects,
  lazy-loaded collections, and metadata that shouldn't leak into the API response.
  Pydantic schemas are pure data shapes: they define exactly what the API sends and
  receives, with validation and serialisation baked in.

  FastAPI uses these schemas to:
    - Validate incoming request bodies (input schemas)
    - Serialise outgoing response bodies (response schemas)
    - Auto-generate the OpenAPI docs at /docs

Naming convention:
  - *Response  — what the API sends to the client
  - *Create    — what the client sends when creating a resource (input)
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, field_validator

from db.status import PropertyStatus

# ── Phase 9 listing-metadata enum-like literals ──────────────────────────────
# These are Pydantic Literals rather than DB-native ENUMs so the same schema
# round-trips through SQLite (tests) and PostgreSQL (production) without a
# dialect-specific ENUM type. Pydantic enforces the allowed value set at the
# API boundary.

ListingType = Literal["rent", "sale"]
PricePeriod = Literal["monthly", "weekly", "nightly", "total"]
PropertyType = Literal["apartment", "house", "villa", "studio", "other"]
Furnishing = Literal["furnished", "semi_furnished", "unfurnished"]
CountryCode = Annotated[str, StringConstraints(min_length=2, max_length=2, to_upper=True)]
NonNegativeDecimal = Annotated[Decimal, Field(ge=Decimal("0"))]

# ── Amenity schemas ──────────────────────────────────────────────────────────


class DetectedAmenityResponse(BaseModel):
    """
    One amenity detection result for a single image.

    Returned as part of PropertyImageResponse, which is part of PropertyResponse.
    """

    id: str
    amenity_name: str
    room_type: str | None
    is_present: bool
    confidence: float | None

    # model_config tells Pydantic to read data from SQLAlchemy ORM attributes.
    # Without this, Pydantic would treat the model as a plain dict and fail to
    # read from ORM objects (which use __getattr__ not __getitem__).
    model_config = ConfigDict(from_attributes=True)


# ── Image schemas ────────────────────────────────────────────────────────────


class PropertyImageResponse(BaseModel):
    """
    One uploaded image for a property, including its detected amenities.
    """

    id: str
    file_path: str
    room_type: str | None
    amenities: list[DetectedAmenityResponse]

    # Phase 9 SEO + ordering fields. alt_text is filled by the same VLM call
    # that detects amenities; is_primary marks the hero image used in JSON-LD.
    alt_text: str | None = None
    caption: str | None = None
    is_primary: bool = False
    display_order: int = 0

    model_config = ConfigDict(from_attributes=True)


# ── Property schemas ─────────────────────────────────────────────────────────


class _PropertyMetadataMixin(BaseModel):
    """Shared Phase 9 fields surfaced on both summary and detail responses.

    Every field is optional so legacy rows (created before Phase 9) load with
    ``None`` and the response stays valid.
    """

    slug: str | None = None
    listing_type: ListingType | None = None
    price: Decimal | None = None
    currency: str | None = None
    price_period: PricePeriod | None = None
    num_bedrooms: int | None = None
    num_bathrooms: int | None = None
    area_sqm: Decimal | None = None
    property_type: PropertyType | None = None
    furnishing: Furnishing | None = None
    available_from: date | None = None
    locality: str | None = None
    postal_code: str | None = None
    country_code: str | None = None
    latitude: Decimal | None = None
    longitude: Decimal | None = None
    owner_email: str | None = None


class PropertySummaryResponse(_PropertyMetadataMixin):
    """
    Lightweight property representation for list endpoints.

    Used in GET /api/v1/properties/ and GET /api/v1/properties/search
    to avoid sending the full amenity list (which can be large).
    """

    id: str
    name: str
    # Publication state (db/status.py). Only 'published' rows reach the public
    # surfaces; the wizard reads it to know whether a draft is already live.
    status: str = PropertyStatus.DRAFT
    description: str | None
    model_used: str | None
    extra_info: str | None
    created_at: datetime
    image_count: int = 0  # Populated from len(property.images) in the router
    first_image_id: str | None = None  # Used by the Gradio browse card thumbnails

    model_config = ConfigDict(from_attributes=True)


class LocalityInsightSummary(BaseModel):
    """The persisted locality enrichment, surfaced on the property detail.

    Read straight off the ``LocalityInsight`` ORM row so the web layer can feed
    it into JSON-LD (geo) + llms.txt and render the neighbourhood panel. ``pois``
    stays as a list of dicts — the raw POIs as stored.
    """

    display_name: str | None = None
    latitude: Decimal | None = None
    longitude: Decimal | None = None
    radius_m: int | None = None
    blurb: str | None = None
    category_counts: dict[str, int] = Field(default_factory=dict)
    transit_breakdown: dict[str, int] | None = None
    pois: list[dict] = Field(default_factory=list)
    attribution: str

    model_config = ConfigDict(from_attributes=True)


class PropertyDetailResponse(_PropertyMetadataMixin):
    """
    Full property details including all images and detected amenities.

    Used in GET /api/v1/properties/{id}.
    """

    id: str
    name: str
    status: str = PropertyStatus.DRAFT
    description: str | None
    model_used: str | None
    extra_info: str | None
    created_at: datetime
    images: list[PropertyImageResponse]
    # Nested neighbourhood enrichment (None until the locality agent has run).
    locality_insight: LocalityInsightSummary | None = None

    model_config = ConfigDict(from_attributes=True)


# ── Upload response schema ────────────────────────────────────────────────────


class PropertyCreateRequest(BaseModel):
    """
    Request body for creating an empty property shell.

    Used by the Phase 5 UI before uploading images one-by-one. Phase 9 added
    the optional listing-metadata fields below; the API stays backward-compatible
    with the original ``{name, model_name, extra_info}`` body.
    """

    name: str
    model_name: str
    extra_info: str | None = None

    # Phase 9 listing metadata. All optional; ``None`` means "not provided" and
    # the corresponding column stays NULL on the row.
    listing_type: ListingType | None = None
    price: NonNegativeDecimal | None = None
    currency: str | None = None
    price_period: PricePeriod | None = None
    num_bedrooms: Annotated[int, Field(ge=0, le=50)] | None = None
    num_bathrooms: Annotated[int, Field(ge=0, le=50)] | None = None
    area_sqm: NonNegativeDecimal | None = None
    property_type: PropertyType | None = None
    furnishing: Furnishing | None = None
    available_from: date | None = None
    locality: str | None = None
    postal_code: str | None = None
    country_code: CountryCode | None = None
    latitude: Annotated[Decimal, Field(ge=-90, le=90)] | None = None
    longitude: Annotated[Decimal, Field(ge=-180, le=180)] | None = None
    # ``owner_email`` is a single-tenant stub today (no auth). Stored as plain
    # str rather than ``EmailStr`` to avoid pulling in the ``email-validator``
    # dependency for what is currently a free-text contact field. When real
    # auth lands in a later phase this will move onto a Users table.
    owner_email: str | None = None


class PropertyUpdateRequest(BaseModel):
    """
    Request body for PATCH /api/v1/properties/{id}.

    Same field set as :class:`PropertyCreateRequest`'s metadata block, minus
    ``name``, ``model_name``, ``extra_info`` (these are creation-time choices)
    and ``slug`` (immutable post-creation to keep public URLs stable).

    ``extra="forbid"`` rejects unknown keys with HTTP 422 — this is what
    blocks attempts to PATCH ``slug``.
    """

    listing_type: ListingType | None = None
    price: NonNegativeDecimal | None = None
    currency: str | None = None
    price_period: PricePeriod | None = None
    num_bedrooms: Annotated[int, Field(ge=0, le=50)] | None = None
    num_bathrooms: Annotated[int, Field(ge=0, le=50)] | None = None
    area_sqm: NonNegativeDecimal | None = None
    property_type: PropertyType | None = None
    furnishing: Furnishing | None = None
    available_from: date | None = None
    locality: str | None = None
    postal_code: str | None = None
    country_code: CountryCode | None = None
    latitude: Annotated[Decimal, Field(ge=-90, le=90)] | None = None
    longitude: Annotated[Decimal, Field(ge=-180, le=180)] | None = None
    owner_email: str | None = None
    # The wizard's describe step calls POST /describe to generate text, then
    # PATCHes that text back here to persist it on the property row.
    description: str | None = None

    model_config = ConfigDict(extra="forbid")


class PropertyCreateResponse(BaseModel):
    """Response returned after creating an empty property shell."""

    property_id: str
    message: str
    property: PropertyDetailResponse


class ImageDetectionResponse(BaseModel):
    """Response returned after processing one image for an existing property."""

    property_id: str
    image: PropertyImageResponse


# ── Health check schema ───────────────────────────────────────────────────────


class HealthResponse(BaseModel):
    """Response from the /health endpoint."""

    status: str
    database: str  # "ok" or "unreachable"
    version: str = "1.0.0"


# ── Image patch schema ───────────────────────────────────────────────────────


class ImageUpdateRequest(BaseModel):
    """
    Request body for PATCH /api/v1/images/{id}.

    All fields are optional; only those actually present in the request are
    written. Setting ``is_primary=True`` flips the hero flag on this image and
    clears the flag on every other image in the same property (one hero per
    property is the contract). ``extra="forbid"`` rejects unknown keys.
    """

    alt_text: str | None = None
    caption: str | None = None
    is_primary: bool | None = None
    display_order: Annotated[int, Field(ge=0, le=999)] | None = None

    model_config = ConfigDict(extra="forbid")


# ── Describe endpoint schemas ─────────────────────────────────────────────────


class AmenityEditItem(BaseModel):
    """One amenity entry as edited by the user in the UI."""

    amenity_name: str
    room_type: str
    is_present: bool


class DescribeRequest(BaseModel):
    """
    Request body for POST /api/v1/properties/{id}/describe.

    The UI sends the user's edited amenity table so the VLM can regenerate
    the description based only on the amenities the user confirmed as present.
    """

    amenities: list[AmenityEditItem]
    model_name: str
    num_rooms: int | None = None
    has_kitchen: bool | None = None
    has_balcony: bool | None = None
    has_living_room: bool | None = None
    hints: dict[str, bool] | None = None


class DescribeResponse(BaseModel):
    """Response from the describe endpoint."""

    description: str


# ── Search endpoint schemas ──────────────────────────────────────────────────


class SearchRequest(BaseModel):
    """Body for ``POST /api/v1/search``.

    A single free-text query — the backend parses it into a structured
    ``SearchFilter`` and runs the hybrid pipeline (SQL filter + cosine
    rerank + FTS blend).
    """

    query: Annotated[str, StringConstraints(min_length=1, max_length=500)]
    limit: Annotated[int, Field(ge=1, le=100)] = 20


# ── Locality enrichment schemas ──────────────────────────────────────────────


class LocalityRequest(BaseModel):
    """Body for the locality endpoints.

    ``postal_code`` (PIN) is required; ``street`` is optional and sharpens the
    centre within the PIN. ``country_code`` (ISO-3166 alpha-2, default DE) keeps
    the PIN from resolving to an identical code in another country. ``radius_m``
    is the everyday-POI search radius (1–10 km), default 3 km; airports are
    searched separately at a fixed wide radius.
    """

    postal_code: Annotated[str, StringConstraints(min_length=1, max_length=16)]
    street: Annotated[str, StringConstraints(max_length=255)] | None = None
    country_code: Annotated[str, StringConstraints(min_length=2, max_length=2)] = "DE"
    radius_m: Annotated[int, Field(ge=1000, le=10000)] = 3000

    @field_validator("country_code")
    @classmethod
    def _upper_country(cls, value: str) -> str:
        return value.upper()


class PoiResponse(BaseModel):
    """A single nearby point of interest."""

    category: str
    name: str
    latitude: float
    longitude: float
    distance_m: float
    osm_type: str
    osm_id: int
    # Transit subtype (bus/tram/subway/light_rail/rail) for transit POIs; else None.
    transit_type: str | None = None


class LocalityResponse(BaseModel):
    """The enriched neighbourhood result returned by both locality endpoints."""

    location_query: str
    display_name: str
    latitude: float | None
    longitude: float | None
    radius_m: int
    blurb: str
    category_counts: dict[str, int]
    transit_breakdown: dict[str, int] | None = None
    pois: list[PoiResponse]
    attribution: str

File: web/lib/schemas.ts
/**
 * Single source of truth for API shapes consumed by the web app.
 *
 * Why Zod instead of bare TS types?
 *   The FastAPI backend returns JSON we cannot blindly trust at runtime — a
 *   shape mismatch (missing field, renamed enum, decimal-as-string vs number)
 *   would otherwise crash a Server Component during render. Zod gives us:
 *     1. Runtime validation at the network boundary.
 *     2. TS types derived from the schema via `z.infer<typeof X>`.
 *   One definition, two guarantees. Keep this file in sync with
 *   `api/schemas.py` — that is the upstream contract.
 *
 * Decimal-as-string note:
 *   Pydantic serialises Decimal as a JSON string ("980" not 980). We coerce
 *   to number at the boundary so React components can format with toLocale*.
 */
import { z } from 'zod';

// ── Phase 9 listing-metadata enums (mirror api/schemas.py Literals) ──────────
export const ListingType = z.enum(['rent', 'sale']);
export type ListingType = z.infer<typeof ListingType>;

export const PricePeriod = z.enum(['monthly', 'weekly', 'nightly', 'total']);
export type PricePeriod = z.infer<typeof PricePeriod>;

export const PropertyType = z.enum(['apartment', 'house', 'villa', 'studio', 'other']);
export type PropertyType = z.infer<typeof PropertyType>;

export const Furnishing = z.enum(['furnished', 'semi_furnished', 'unfurnished']);
export type Furnishing = z.infer<typeof Furnishing>;

/** ISO-3166 alpha-2, normalised to upper-case to match FastAPI StringConstraints. */
const CountryCode = z
  .string()
  .length(2)
  .transform((s) => s.toUpperCase());

/** Pydantic Decimal arrives as a JSON string. Coerce to number on the client. */
const DecimalLike = z.union([z.number(), z.string()]).transform((v) => Number(v));

const Latitude = DecimalLike.refine((n) => n >= -90 && n <= 90, 'latitude out of range');
const Longitude = DecimalLike.refine((n) => n >= -180 && n <= 180, 'longitude out of range');

// ── Detected amenity ─────────────────────────────────────────────────────────
export const DetectedAmenity = z.object({
  id: z.string(),
  amenity_name: z.string().min(1),
  room_type: z.string().nullable(),
  is_present: z.boolean(),
  /** 0..1 confidence from the VLM, or null if the user added the amenity by hand. */
  confidence: z.number().nullable(),
});
export type DetectedAmenity = z.infer<typeof DetectedAmenity>;

// ── Property image ───────────────────────────────────────────────────────────
export const PropertyImage = z.object({
  id: z.string(),
  file_path: z.string(),
  room_type: z.string().nullable(),
  amenities: z.array(DetectedAmenity),
  alt_text: z.string().nullable().default(null),
  caption: z.string().nullable().default(null),
  is_primary: z.boolean().default(false),
  display_order: z.number().int().nonnegative().default(0),
});
export type PropertyImage = z.infer<typeof PropertyImage>;

// ── Shared Phase 9 listing metadata block ────────────────────────────────────
const ListingMetadata = z.object({
  slug: z.string().nullable().default(null),
  listing_type: ListingType.nullable().default(null),
  price: DecimalLike.nullable().default(null),
  currency: z.string().nullable().default(null),
  price_period: PricePeriod.nullable().default(null),
  num_bedrooms: z.number().int().nullable().default(null),
  num_bathrooms: z.number().int().nullable().default(null),
  area_sqm: DecimalLike.nullable().default(null),
  property_type: PropertyType.nullable().default(null),
  furnishing: Furnishing.nullable().default(null),
  available_from: z.string().nullable().default(null),
  locality: z.string().nullable().default(null),
  postal_code: z.string().nullable().default(null),
  country_code: z.string().nullable().default(null),
  latitude: Latitude.nullable().default(null),
  longitude: Longitude.nullable().default(null),
  owner_email: z.string().nullable().default(null),
});

// ── PropertySummary (list endpoints) ─────────────────────────────────────────
export const PropertySummary = ListingMetadata.extend({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  model_used: z.string().nullable(),
  extra_info: z.string().nullable(),
  created_at: z.string(),
  image_count: z.number().int().nonnegative().default(0),
  first_image_id: z.string().nullable().default(null),
});
export type PropertySummary = z.infer<typeof PropertySummary>;

// ── Locality insight (nested on PropertyDetail) ──────────────────────────────
// Mirrors api/schemas.py LocalityInsightSummary. The neighbourhood enrichment
// produced by the locality agent: a written "Lage" blurb, per-category POI
// counts, the raw nearby POIs, and the mandatory OSM attribution.
export const LocalityPoi = z.object({
  category: z.string(),
  name: z.string().default(''),
  latitude: z.number(),
  longitude: z.number(),
  distance_m: z.number().default(0),
  osm_type: z.string().default(''),
  osm_id: z.number().default(0),
  // Transit subtype for the "transit" category: bus/tram/subway/light_rail/rail.
  // null/absent for every other category (and for stops we couldn't classify,
  // including legacy rows persisted before subtypes existed).
  transit_type: z.string().nullish(),
});
export type LocalityPoi = z.infer<typeof LocalityPoi>;

// Request body for the locality endpoints (mirrors api/schemas.py LocalityRequest).
// PIN required; street optional; country defaults to DE; everyday radius 1–10 km.
export const LocalityRequest = z.object({
  postal_code: z.string().min(1).max(16),
  street: z.string().max(255).optional(),
  country_code: CountryCode.default('DE'),
  radius_m: z.number().int().min(1000).max(10000).default(3000),
});
export type LocalityRequest = z.infer<typeof LocalityRequest>;

export const LocalityInsight = z.object({
  display_name: z.string().nullable().default(null),
  latitude: Latitude.nullable().default(null),
  longitude: Longitude.nullable().default(null),
  radius_m: z.number().int().nullable().default(null),
  blurb: z.string().nullable().default(null),
  category_counts: z.record(z.string(), z.number()).default({}),
  pois: z.array(LocalityPoi).default([]),
  attribution: z.string(),
});
export type LocalityInsight = z.infer<typeof LocalityInsight>;

// ── PropertyDetail (single property endpoint) ────────────────────────────────
export const PropertyDetail = ListingMetadata.extend({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  model_used: z.string().nullable(),
  extra_info: z.string().nullable(),
  created_at: z.string(),
  images: z.array(PropertyImage),
  // Optional so existing fixtures/legacy rows without enrichment still parse.
  locality_insight: LocalityInsight.nullable().optional(),
});
export type PropertyDetail = z.infer<typeof PropertyDetail>;

// ── Create / update request bodies ───────────────────────────────────────────
// These mirror api/schemas.py PropertyCreateRequest / PropertyUpdateRequest.
// We re-validate on the client so we never POST a body that the server will
// 422 back at us.

const _listingFields = {
  listing_type: ListingType.optional(),
  price: z.number().nonnegative().optional(),
  currency: z.string().optional(),
  price_period: PricePeriod.optional(),
  num_bedrooms: z.number().int().min(0).max(50).optional(),
  num_bathrooms: z.number().int().min(0).max(50).optional(),
  area_sqm: z.number().nonnegative().optional(),
  property_type: PropertyType.optional(),
  furnishing: Furnishing.optional(),
  available_from: z.string().optional(),
  locality: z.string().optional(),
  postal_code: z.string().optional(),
  country_code: CountryCode.optional(),
  latitude: z.number().min(-90).max(90).optional(),
  longitude: z.number().min(-180).max(180).optional(),
  owner_email: z.string().optional(),
} as const;

export const PropertyCreateRequest = z.object({
  name: z.string().min(1),
  model_name: z.string().min(1),
  extra_info: z.string().optional(),
  ..._listingFields,
});
export type PropertyCreateRequest = z.infer<typeof PropertyCreateRequest>;

export const PropertyUpdateRequest = z
  .object({
    ..._listingFields,
    // The wizard PATCHes the generated description here after the describe
    // step. Backend mirrors this field on api/schemas.py PropertyUpdateRequest.
    description: z.string().optional(),
  })
  .strict();
export type PropertyUpdateRequest = z.infer<typeof PropertyUpdateRequest>;

// ── Image PATCH body ─────────────────────────────────────────────────────────
export const ImageUpdateRequest = z
  .object({
    alt_text: z.string().optional(),
    caption: z.string().optional(),
    is_primary: z.boolean().optional(),
    display_order: z.number().int().min(0).max(999).optional(),
  })
  .strict();
export type ImageUpdateRequest = z.infer<typeof ImageUpdateRequest>;

// ── Describe endpoint ────────────────────────────────────────────────────────
export const AmenityEditItem = z.object({
  amenity_name: z.string(),
  room_type: z.string(),
  is_present: z.boolean(),
});
export type AmenityEditItem = z.infer<typeof AmenityEditItem>;

export const DescribeRequest = z.object({
  amenities: z.array(AmenityEditItem),
  model_name: z.string(),
  num_rooms: z.number().int().optional(),
  has_kitchen: z.boolean().optional(),
  has_balcony: z.boolean().optional(),
  has_living_room: z.boolean().optional(),
  hints: z.record(z.string(), z.boolean()).optional(),
});
export type DescribeRequest = z.infer<typeof DescribeRequest>;

export const DescribeResponse = z.object({ description: z.string() });
export type DescribeResponse = z.infer<typeof DescribeResponse>;

// ── Mutation response wrappers ───────────────────────────────────────────────
// These mirror api/schemas.py PropertyCreateResponse and ImageDetectionResponse.

export const PropertyCreateResponse = z.object({
  property_id: z.string(),
  message: z.string(),
  property: PropertyDetail,
});
export type PropertyCreateResponse = z.infer<typeof PropertyCreateResponse>;

export const ImageDetectionResponse = z.object({
  property_id: z.string(),
  image: PropertyImage,
});
export type ImageDetectionResponse = z.infer<typeof ImageDetectionResponse>;

// ── Model registry info ──────────────────────────────────────────────────────
// Returned by GET /api/v1/models/ — feeds the wizard's model dropdown.

export const ModelInfo = z.object({
  name: z.string(),
  available: z.boolean(),
  description: z.string(),
});
export type ModelInfo = z.infer<typeof ModelInfo>;

// ── Health check ─────────────────────────────────────────────────────────────
export const Health = z.object({
  status: z.string(),
  database: z.string(),
  version: z.string().default('1.0.0'),
});
export type Health = z.infer<typeof Health>;

// ── Client-only: natural-language search filter ──────────────────────────────
// Lives here even though FastAPI does not own it, because the URL/search-bar
// is the source of truth for the browse page. Phase 11 backend will consume
// the same shape via /api/v1/properties/search.
export const SearchFilter = z.object({
  query: z.string().default(''),
  rooms: z.union([z.literal('any'), z.number().int().min(1).max(5)]).default('any'),
  amenities: z.array(z.string()).default([]),
  sort: z.enum(['recent', 'mostAmenities', 'name']).default('recent'),
});
export type SearchFilter = z.infer<typeof SearchFilter>;
